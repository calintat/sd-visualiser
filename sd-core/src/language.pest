variable = @{ !("bind" | "in") ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }
bind_clause = { "bind" ~ variable ~ "=" ~ term ~ "in" }
expr = { bind_clause* ~ value }
term = { thunk | (active_op ~ "(" ~ args ~ ")") | value }
value = { (passive_op ~ "(" ~ args ~ ")") | variable }
active_op = { "plus" | "minus" | "times" | "eq" | "if" | "app" | "lambda" | "rec" }
passive_op = { number }
number = _{ ASCII_DIGIT+ }
args = _{ value ~ ("," ~ value)* | "" }
program = _{ SOI ~ expr ~ EOI }
WHITESPACE = _{ " " | "\t" | "\n" }
thunk = { variable* ~ "." ~ expr }