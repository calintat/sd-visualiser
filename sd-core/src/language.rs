pub mod chil;
pub mod spartan;

pub(super) mod visitor;
