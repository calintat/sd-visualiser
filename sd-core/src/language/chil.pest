program = _{ SOI ~ expr ~ EOI }

keyword = { "def" | "output" | "thunk" }

expr = { bind_clause* ~ "output" ~ value }
bind_clause = { "def" ~ variable_def ~ "=" ~ value }

thunk = { "thunk" ~ addr ~ "=" ~ "{" ~ (variable_def ~ ("," ~ variable_def)*)? ~ "=>" ~ expr ~ "}" }

value = { variable | op ~ ("(" ~ (thunk+ | values ~ (";" ~ thunk+)?) ~ ")")? }
values = _{ value ~ ("," ~ value)* }

op = ${ !keyword ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "+" | "-" | "*" | "=")* ~ ("/" ~ (number | string | ty ~ "/" ~ ty))? }
number = @{ ASCII_DIGIT+ ~ ("." ~ ASCII_DIGIT+)? }
string = { "\"" ~ (!"\"" ~ ANY)* ~ "\"" }

ty = { generic_ty | base_ty | function_ty | tuple_ty }
tys = _{ ty ~ ("," ~ ty)* }
base_ty = @{ 'A'..'Z' ~ (ASCII_ALPHANUMERIC | "-")* }
generic_ty = { base_ty ~ "<" ~ tys ~ ">" }
tuple_ty = { "(" ~ tys? ~ ")" }
function_ty = { tuple_ty ~ "->" ~ ty }

variable = { addr | identifier ~ "(" ~ "id" ~ ":" ~ addr ~ ")" }
variable_def = { variable ~ ":" ~ ty | variable }

COMMENT = _{ "#" ~ (!"\n" ~ ANY)* }
WHITESPACE = _{ " " | "\t" | "\n" }

addr = @{ ("@" | "%") ~ ASCII_DIGIT+ }
identifier = @{ !keyword ~ "_"* ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }