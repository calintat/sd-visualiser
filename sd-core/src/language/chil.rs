#![allow(clippy::clone_on_copy)]

use std::fmt::{Display, Write};

use from_pest::{ConversionError, FromPest, Void};
use pest::iterators::Pairs;
use pest_ast::FromPest;
use pest_derive::Parser;

use super::{span_into_str, AsVar};

pub struct Chil;

impl super::Language for Chil {
    type Op = Op;
    type Var = Variable;
    type Addr = Addr;
    type VarDef = VariableDef;

    type Rule = Rule;

    fn expr_rule() -> Self::Rule {
        Rule::expr
    }
    fn bind_rule() -> Self::Rule {
        Rule::bind
    }
    fn value_rule() -> Self::Rule {
        Rule::value
    }
    fn thunk_rule() -> Self::Rule {
        Rule::thunk
    }
}

pub type Expr = super::Expr<Chil>;
pub type Bind = super::Bind<Chil>;
pub type Value = super::Value<Chil>;
pub type Thunk = super::Thunk<Chil>;

#[derive(Parser)]
#[grammar = "language/chil.pest"]
pub struct ChilParser;

fn parse_addr_first(input: &str) -> char {
    input.chars().next().unwrap()
}

fn parse_addr_second(input: &str) -> usize {
    input[1..].parse().unwrap()
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct Op(pub String);

impl Display for Op {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.0.as_str() {
            "+" | "throwing+" => f.write_char('+'),
            "-" | "throwing-" => f.write_char('-'),
            "*" | "throwing*" => f.write_char('*'),
            "/" | "throwing/" => f.write_char('/'),
            "%" | "throwing%" => f.write_char('%'),
            "&&" | "throwing&&" => f.write_str("&&"),
            "||" | "throwing||" => f.write_str("||"),
            "!" | "throwing!" => f.write_char('!'),
            "==" | "throwing==" => f.write_char('='),
            "!=" | "throwing!=" => f.write_char('≠'),
            "<" | "throwing<" => f.write_char('<'),
            "<=" | "throwing<=" => f.write_char('≤'),
            ">" | "throwing>" => f.write_char('>'),
            ">=" | "throwing>=" => f.write_char('≥'),
            "func" => f.write_char('λ'),
            "seq" => f.write_char(';'),
            "unit" => f.write_str("()"),
            str => {
                if str.starts_with("apply/") {
                    return f.write_char('@');
                }
                if str.starts_with("tuple/") {
                    return f.write_char('⊗');
                }
                if let Some(rest) = None
                    .or_else(|| str.strip_prefix("bool/"))
                    .or_else(|| str.strip_prefix("int64/"))
                    .or_else(|| str.strip_prefix("float64/"))
                    .or_else(|| str.strip_prefix("string/"))
                {
                    return f.write_str(rest);
                }
                f.write_str(str)
            }
        }
    }
}

impl<'pest> FromPest<'pest> for Op {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::op {
            return Err(ConversionError::NoMatch);
        }
        *pest = clone;
        Ok(Self(pair.as_str().to_owned()))
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::variable))]
pub struct Variable {
    pub name: Option<Identifier>,
    pub addr: Addr,
}

impl Display for Variable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.name {
            None => write!(f, "{}", self.addr),
            Some(name) => write!(f, "{}(id: {})", name, self.addr),
        }
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::addr))]
pub struct Addr(
    #[pest_ast(outer(with(span_into_str), with(parse_addr_first)))] pub char,
    #[pest_ast(outer(with(span_into_str), with(parse_addr_second)))] pub usize,
);

impl Display for Addr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", self.0, self.1)
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct Identifier(pub String);

impl Display for Identifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl<'pest> FromPest<'pest> for Identifier {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::identifier {
            return Err(ConversionError::NoMatch);
        }
        *pest = clone;
        Ok(Self(pair.as_str().to_owned()))
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::variable_def))]
pub struct VariableDef {
    pub var: Variable,
    pub r#type: Option<Type>,
}

impl AsVar<Variable> for VariableDef {
    fn as_var(&self) -> &Variable {
        &self.var
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::ty))]
pub enum Type {
    Base(BaseType),
    Generic(GenericType),
    Tuple(TupleType),
    Function(FunctionType),
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::base_ty))]
pub struct BaseType(#[pest_ast(outer(with(span_into_str), with(str::to_string)))] pub String);

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::generic_ty))]
pub struct GenericType {
    pub base: BaseType,
    pub params: Vec<Type>,
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::tuple_ty))]
pub struct TupleType {
    pub types: Vec<Type>,
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::function_ty))]
pub struct FunctionType {
    pub domain: TupleType,
    pub codomain: Box<Type>,
}

// Conversion to spartan

#[cfg(test)]
impl From<Op> for super::spartan::Op {
    fn from(op: Op) -> Self {
        match op.0.as_str() {
            "+" | "throwing+" => return Self::Plus,
            "-" | "throwing-" => return Self::Minus,
            "*" | "throwing*" => return Self::Times,
            "/" | "throwing/" => return Self::Div,
            "%" | "throwing%" => return Self::Rem,
            "&&" | "throwing&&" => return Self::And,
            "||" | "throwing||" => return Self::Or,
            "!" | "throwing!" => return Self::Not,
            "==" | "throwing==" => return Self::Eq,
            "!=" | "throwing!=" => return Self::Neq,
            "<" | "throwing<" => return Self::Lt,
            "<=" | "throwing<=" => return Self::Leq,
            ">" | "throwing>" => return Self::Gt,
            ">=" | "throwing>=" => return Self::Geq,
            "func" => return Self::Lambda,
            "atom" => return Self::Atom,
            "deref" => return Self::Deref,
            "asg" => return Self::Assign,
            "bool/true" => return Self::Bool(true),
            "bool/false" => return Self::Bool(false),
            str => {
                if str.starts_with("apply/") {
                    return Self::App;
                }
                if let Some(rest) = str.strip_prefix("int64/") {
                    return Self::Number(rest.parse().unwrap_or_default());
                }
                Self::Plus // dummy placeholder
            }
        }
    }
}

#[cfg(test)]
impl From<Variable> for super::spartan::Variable {
    fn from(var: Variable) -> Self {
        Self(format!("var_{}", var.addr.1))
    }
}

#[cfg(test)]
impl From<Addr> for super::spartan::Addr {
    fn from(_addr: Addr) -> Self {
        Self
    }
}

#[cfg(test)]
impl From<VariableDef> for super::spartan::Variable {
    fn from(def: VariableDef) -> Self {
        def.var.into()
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::path::Path;

    use dir_test::{dir_test, Fixture};
    use from_pest::FromPest;
    use pest::Parser;

    use super::{ChilParser, Expr, Rule};

    pub fn parse_chil(raw_path: &str) -> (&str, Expr) {
        let path = Path::new(raw_path);
        let program = std::fs::read_to_string(path).unwrap();
        let mut pairs = ChilParser::parse(Rule::program, &program).unwrap_or_else(|err| {
            panic!(
                "could not parse program {:?}\n{err:?}",
                path.file_stem().unwrap()
            )
        });
        let name = path.file_stem().unwrap().to_str().unwrap();
        let expr = Expr::from_pest(&mut pairs).unwrap();
        (name, expr)
    }

    #[allow(clippy::needless_pass_by_value)]
    #[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/*.chil", loader: crate::language::chil::tests::parse_chil, postfix: "check_parse")]
    fn check_parse(fixture: Fixture<(&str, Expr)>) {
        let (_name, _expr) = fixture.content();
    }
}
