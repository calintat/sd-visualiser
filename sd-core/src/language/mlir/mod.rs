#![allow(clippy::clone_on_copy)]

use std::fmt::Display;

use itertools::Itertools;

pub mod internal;

#[cfg(test)]
use serde::Serialize;

use self::internal::Attribute;
use super::{Fresh, Language, OpInfo, CF};
use crate::common::{Matchable, Unit};

pub struct Mlir;

impl Language for Mlir {
    type Op = Op;
    type Var = Var;
    type Addr = Unit;
    type VarDef = Var;
    type BlockAddr = BlockAddr;
}

pub type Expr = super::Expr<Mlir>;
pub type Bind = super::Bind<Mlir>;
pub type Value = super::Value<Mlir>;
pub type Thunk = super::Thunk<Mlir>;
pub type Block = super::Block<Mlir>;

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct Op {
    pub name: String,
    pub successors: Vec<BlockAddr>,
    pub attributes: String,
    pub sym_name: Option<String>,
    pub symbols: Vec<String>,
}

impl Display for Op {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.name)
    }
}

impl Matchable for Op {
    fn is_match(&self, query: &str) -> bool {
        self.name == query
    }
}

impl OpInfo<Mlir> for Op {
    fn get_cf(&self) -> Option<CF<Mlir>> {
        if !self.successors.is_empty() {
            Some(CF::Brs(self.successors.clone()))
        } else if self.name.contains("return") || self.name == "arc.output" || self.name == "yield"
        {
            Some(CF::Return)
        } else {
            None
        }
    }

    fn symbol_use(&self) -> impl Iterator<Item = &str> {
        self.symbols.iter().map(|x| x.as_str())
    }

    fn sym_name(&self) -> Option<&str> {
        self.sym_name.as_deref()
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct Var {
    pub id: String,
    pub index: Option<usize>,
}

impl Display for Var {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(idx) = self.index {
            write!(f, "{}#{idx}", self.id)
        } else {
            f.write_str(&self.id)
        }
    }
}

impl Matchable for Var {
    fn is_match(&self, query: &str) -> bool {
        self.id == query
    }
}

impl Fresh for Var {
    fn fresh(number: usize) -> Self {
        Var {
            id: format!("?{number}"),
            index: None,
        }
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct BlockAddr(pub String);

impl Display for BlockAddr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl Matchable for BlockAddr {
    fn is_match(&self, query: &str) -> bool {
        self.0 == query
    }
}

// Conversion from internal AST.

impl From<internal::Value> for Var {
    fn from(value: internal::Value) -> Self {
        Var {
            id: value.id,
            index: value.index.map(|idx| idx.0),
        }
    }
}

impl From<internal::TypedArg> for Var {
    fn from(arg: internal::TypedArg) -> Self {
        Var {
            id: arg.id,
            index: None,
        }
    }
}

impl From<internal::OpResult> for Vec<Var> {
    fn from(op_result: internal::OpResult) -> Vec<Var> {
        if let Some(idx) = op_result.index {
            (0..idx.0)
                .map(|x| Var {
                    id: op_result.id.clone(),
                    index: Some(x),
                })
                .collect()
        } else {
            vec![Var {
                id: op_result.id,
                index: None,
            }]
        }
    }
}

impl From<internal::BlockId> for BlockAddr {
    fn from(id: internal::BlockId) -> Self {
        BlockAddr(id.0)
    }
}

impl From<internal::Successor> for BlockAddr {
    fn from(successor: internal::Successor) -> Self {
        BlockAddr(successor.id)
    }
}

impl From<Vec<internal::Operation>> for Expr {
    fn from(ops: Vec<internal::Operation>) -> Self {
        Expr {
            binds: ops.into_iter().map_into().collect(),
            values: vec![],
        }
    }
}

impl From<internal::Operation> for Bind {
    fn from(op: internal::Operation) -> Self {
        Bind {
            defs: op.result.into_iter().map_into::<Vec<Var>>().concat(),
            value: op.operation.into(),
        }
    }
}

impl From<internal::Region> for Thunk {
    fn from(region: internal::Region) -> Self {
        Thunk {
            addr: Unit,
            args: vec![],
            body: region
                .entry_block
                .map(|block| block.operations.into())
                .unwrap_or_default(),
            blocks: region.blocks.into_iter().map_into().collect(),
        }
    }
}

impl From<internal::Block> for Block {
    fn from(block: internal::Block) -> Self {
        Block {
            addr: block.label.id.into(),
            args: block.label.args.into_iter().map_into().collect(),
            expr: block.operations.into(),
        }
    }
}

impl From<internal::GenericOperation> for Value {
    fn from(generic_op: internal::GenericOperation) -> Self {
        Value::Op {
            op: Op {
                name: generic_op.op,
                successors: generic_op.successors.into_iter().map_into().collect(),
                attributes: if generic_op.attributes.len() + generic_op.properties.len() == 0 {
                    String::new()
                } else {
                    format!(
                        "{{{}}}",
                        generic_op
                            .attributes
                            .iter()
                            .chain(generic_op.properties.iter())
                            .join(", ")
                    )
                },
                sym_name: generic_op
                    .attributes
                    .iter()
                    .chain(generic_op.properties.iter())
                    .map(Attribute::is_sym_name)
                    .find(Option::is_some)
                    .flatten(),
                symbols: generic_op
                    .attributes
                    .iter()
                    .chain(generic_op.properties.iter())
                    .filter_map(|attr| attr.get_symbol())
                    .collect(),
            },
            args: generic_op
                .operands
                .into_iter()
                .map(|x| Value::Variable(x.into()))
                .chain(
                    generic_op
                        .regions
                        .into_iter()
                        .map(|x| Value::Thunk(x.into())),
                )
                .collect(),
        }
    }
}
