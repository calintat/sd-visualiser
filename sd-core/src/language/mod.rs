use std::{
    fmt::{Debug, Display},
    hash::Hash,
};

use derivative::Derivative;

use crate::{common::Matchable, prettyprinter::PrettyPrint};

pub mod chil;
pub mod spartan;

pub(crate) fn span_into_str(span: pest::Span) -> &str {
    span.as_str()
}

pub trait GetVar<V> {
    fn var(&self) -> &V;
    fn into_var(self) -> V;
}

impl<V> GetVar<V> for V {
    fn var(&self) -> &V {
        self
    }

    fn into_var(self) -> V {
        self
    }
}

pub trait Fresh {
    fn fresh(number: usize) -> Self;
}

/// `Display` should give the symbolic representation (e.g. "+").
/// `PrettyPrint` should give the textual representation (e.g. "plus").
pub trait Syntax:
    Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint
{
}
impl<T: Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint> Syntax for T {}

pub struct BlockWithArgs<T: Language + ?Sized> {
    pub block: T::BlockAddr,
    pub args: Vec<Value<T>>,
}

pub enum CF<T: Language + ?Sized> {
    Return,
    Brs(Vec<BlockWithArgs<T>>),
}

pub trait ControlFlow<T: Language + ?Sized> {
    fn get_cf(&self) -> Option<CF<T>>;

    #[must_use]
    fn get_apply() -> Option<T::Op> {
        None
    }
}

pub trait Language {
    type Op: Syntax + ControlFlow<Self>;
    type Var: Syntax + Fresh;
    type Addr: Syntax;
    type BlockAddr: Syntax;
    type VarDef: Syntax + GetVar<Self::Var>;
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Expr<T: Language + ?Sized> {
    pub binds: Vec<Bind<T>>,
    pub values: Vec<Value<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Bind<T: Language + ?Sized> {
    pub defs: Vec<T::VarDef>,
    pub value: Value<T>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub enum Value<T: Language + ?Sized> {
    Variable(T::Var),
    Thunk(Thunk<T>),
    Op { op: T::Op, args: Vec<Value<T>> },
}

impl<T: Language> ControlFlow<T> for Value<T> {
    fn get_cf(&self) -> Option<CF<T>> {
        match self {
            Value::Op { op, .. } => op.get_cf(),
            _ => None,
        }
    }
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Thunk<T: Language + ?Sized> {
    pub addr: T::Addr,
    pub args: Vec<T::VarDef>,
    pub body: Expr<T>,
    pub blocks: Vec<Block<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Block<T: Language + ?Sized> {
    pub addr: T::BlockAddr,
    pub args: Vec<T::VarDef>,
    pub expr: Expr<T>,
}

// Conversions between languages

impl<T: Language> Expr<T> {
    pub fn into<U: Language>(self) -> Expr<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Expr {
            binds: self.binds.into_iter().map(Bind::into).collect(),
            values: self.values.into_iter().map(Value::into).collect(),
        }
    }
}

impl<T: Language> Bind<T> {
    pub fn into<U: Language>(self) -> Bind<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Bind {
            defs: self.defs.into_iter().map(Into::into).collect(),
            value: self.value.into(),
        }
    }
}

impl<T: Language> Value<T> {
    pub fn into<U: Language>(self) -> Value<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        match self {
            Self::Variable(var) => Value::Variable(var.into()),
            Self::Thunk(thunk) => Value::Thunk(thunk.into()),
            Self::Op { op, args } => Value::Op {
                op: op.into(),
                args: args.into_iter().map(Value::into).collect(),
            },
        }
    }
}

impl<T: Language> Thunk<T> {
    pub fn into<U: Language>(self) -> Thunk<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Thunk {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            body: self.body.into(),
            blocks: self.blocks.into_iter().map(Block::into).collect(),
        }
    }
}

impl<T: Language> Block<T> {
    pub fn into<U: Language>(self) -> Block<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Block {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            expr: self.expr.into(),
        }
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::{ffi::OsStr, path::Path};

    use super::{
        chil::tests::parse_chil,
        spartan::{self, tests::parse_sd},
    };

    pub fn parse(raw_path: &str) -> (&str, &str, spartan::Expr) {
        let path = Path::new(raw_path);
        match path.extension() {
            Some(ext) if ext == OsStr::new("sd") => {
                let (name, expr) = parse_sd(raw_path);
                ("sd", name, expr)
            }
            Some(ext) if ext == OsStr::new("chil") => {
                let (name, expr) = parse_chil(raw_path);
                ("chil", name, expr.into())
            }
            _ => unreachable!(),
        }
    }
}
