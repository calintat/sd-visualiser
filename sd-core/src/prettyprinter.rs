use crate::language::{ActiveOp, BindClause, Expr, PassiveOp, Term, Thunk, Value, Variable};
use pretty::RcDoc;

pub trait PrettyPrint: std::fmt::Debug {
    fn to_doc(&self) -> RcDoc<'_, ()>;
    #[tracing::instrument]
    fn to_pretty(&self) -> String {
        tracing::debug!("self.to_doc(): {:#?}", self.to_doc());
        self.to_doc()
            .pretty(usize::MAX) // in principle, the pretty printer could do different things at different widths
            .to_string()
    }
}

impl PrettyPrint for Expr {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::concat(self.binds.iter().map(|bind| bind.to_doc())).append(self.value.to_doc())
    }
}

impl PrettyPrint for BindClause {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text("bind")
            .append(RcDoc::space())
            .append(self.var.to_doc())
            .append(RcDoc::space())
            .append(RcDoc::text("="))
            .append(RcDoc::space())
            .append(self.term.to_doc())
            // term will have either a space or \n (for thunks) at the end
            .append(RcDoc::text("in"))
            .append(RcDoc::line())
    }
}

impl PrettyPrint for Variable {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(self.0.to_owned())
    }
}

impl PrettyPrint for Term {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Term::Thunk(thunk) => thunk.to_doc().append(if thunk.body.binds.is_empty() {
                RcDoc::space()
            } else {
                RcDoc::line()
            }),
            Term::ActiveOp(op, vs) => op
                .to_doc()
                .append(RcDoc::text("("))
                .append(RcDoc::intersperse(
                    vs.iter().map(|v| v.to_doc()),
                    RcDoc::text(",").append(RcDoc::space()),
                ))
                .append(RcDoc::text(")"))
                .append(RcDoc::space()),
            Term::Value(v) => v.to_doc().append(RcDoc::space()),
        }
    }
}

impl PrettyPrint for ActiveOp {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(match *self {
            ActiveOp::Plus => "plus",
            ActiveOp::Minus => "minus",
            ActiveOp::Times => "times",
            ActiveOp::Eq => "eq",
            ActiveOp::And => "and",
            ActiveOp::Or => "or",
            ActiveOp::Not => "not",
            ActiveOp::If => "if",
            ActiveOp::App => "app",
            ActiveOp::Lambda => "lambda",
            ActiveOp::Rec => "rec",
        })
    }
}

impl PrettyPrint for PassiveOp {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(match *self {
            PassiveOp::Int(n) => n.to_string(),
            PassiveOp::Bool(b) => b.to_string(),
        })
    }
}

impl PrettyPrint for Value {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Value::PassiveOp(op, vs) => {
                if vs.is_empty() {
                    op.to_doc()
                } else {
                    op.to_doc()
                        .append(RcDoc::text("("))
                        .append(RcDoc::intersperse(
                            vs.iter().map(|v| v.to_doc()),
                            RcDoc::text(",").append(RcDoc::space()),
                        ))
                        .append(RcDoc::text(")"))
                }
            }
            Value::Var(v) => v.to_doc(),
        }
    }
}

impl PrettyPrint for Thunk {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::intersperse(self.args.iter().map(|arg| arg.to_doc()), RcDoc::space())
            .append(RcDoc::space())
            .append(RcDoc::text("."))
            .append(if self.body.binds.is_empty() {
                RcDoc::space().append(self.body.to_doc())
            } else {
                RcDoc::line().append(self.body.to_doc()).nest(4)
            })
    }
}

#[cfg(test)]
mod tests {
    use crate::language::tests::*;
    use anyhow::Result;
    use insta::assert_snapshot;
    use rstest::rstest;

    use crate::{language::Expr, prettyprinter::PrettyPrint};

    #[rstest]
    #[case("basic_program", basic_program())]
    #[case("free_vars", free_vars())]
    #[case("thunks", thunks())]
    #[case("fact", fact())]
    fn print_snapshots(#[case] name: &str, #[case] expr: Result<Expr>) -> Result<()> {
        let expr = expr?;
        // have to manually specify name due to https://github.com/la10736/rstest/issues/183
        assert_snapshot!(name, expr.to_pretty());
        Ok(())
    }
}
