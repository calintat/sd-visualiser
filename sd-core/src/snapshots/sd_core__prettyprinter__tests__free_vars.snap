---
source: sd-core/src/prettyprinter.rs
expression: expr.to_pretty()
---
bind x = y in
z
